### ISSUE TITLE
Requirements for running user space singularity

### ISSUE THREAD

**Issue by tmefford** [apptainer/singularity#1468]
## Version of Singularity:
2.4.5

### Expected behavior
Install in user space

### Actual behavior

Further configuration required..

### Steps to reproduce behavior

On page:  http://singularity.lbl.gov/docs-config
 
In the section on “ALLOW SETUID” it says:
 
“You can however disable SUID support iff (if and only if) you do not need to use the default Singularity image file format and if your kernel supports user namespaces and you choose to use user namespaces.”
 
We are discussing getting Singularity running on an HPC compute center.   They seem unwilling to install Singularity as root, but allowed the possibility of enabling user namespace.
 
We are not yet singularity experts, though, and the above comment isn’t entirely clear to us, namely “default Singularity image file format”
 
What other image file format would we use?    Is there an example anywhere of using the default and then of using something else?   Sorry, but I don't quite follow that comment.
 
Any light you could throw on the issue would be appreciated.

**Comment by dtrudg** [comment]
Hi @tmefford - you would need to use the `-s` sandbox option to Singularity build, which creates a container inside a directory, rather than as a single image file. So to build a container then run it, your commands would be similar to this:

```
singularity build -s mysandbox docker://alpine

singularity run --userns mysandbox/
```

Without the -s sandbox option by default you would be building a single file holding the container, which would be in 'squashfs' format.

Apologies the documentation isn't clearer on this - it needs to be expanded there.

**Comment by tmefford** [comment]
Thank you, David,

       If I wanted to run a singularity container I’d built elsewhere, That container would have to be built the same way, and then move the whole directory to the other system?

Thank you,

Tim

**Comment by dtrudg** [comment]
Sorry for the slow reply. Yes you'd need to build it into a sandbox with -s on the other machine, then copy it to the location you want to run it. Be careful that the method you use to copy preserves permissions and timestamps etc. to avoid strange behaviour.

**Comment by ghost** [comment]
The documentation says somewhat like "Singularity needs user namespaces or SUID permissions to execute suid portions of code, if exists".

Then it says if run without SUID, many features will fail && user namespaces wasn't that robust (at least at the time of writing).

Is there any solution (At least a hack) to run Singularity without (user namespaces && suid) i.e without any admin intervention. Like unpack the tarball somewhere, grab the images and go.

**Comment by dtrudg** [comment]
We had considerable updates in the 3.5 documentation to clarify user namespace functionality.

https://sylabs.io/guides/3.5/admin-guide/user_namespace.html

> Is there any solution (At least a hack) to run Singularity without (user namespaces && suid) i.e without any admin intervention. Like unpack the tarball somewhere, grab the images and go.

You need user namespaces available, or to make setuid installation. The build process for non-setuid is now documented https://sylabs.io/guides/3.5/admin-guide/installation.html#install-nonsetuid

**Comment by dmey** [comment]
@dtrudg when building without SUID, Singularity converts the SIF file to a temporary sandbox. Aside from having to do this every time the image is run, thus slowing down pipelines considerably when multiple images need to be run in succession, I also noticed a large amount of memory being used (5 GB each time). This takes a while before being garbage collected thus leading to further slowdowns when all memory has been used up. Is this behaviour expected? If so, would it be worth mentioning it in the docs? I tried on 3.7.3 but noticed the same with other versions.
